ONSdigital/dp-feedback-api
# Implement POST /feedback

### What
[Trello card](https://trello.com/c/j8pKkXGy/1139-create-post-feedback-endpoint)

- refactored service package to have cleaner code according to lifecycle (init and start)
  - please see [another API repository](https://github.com/ONSdigital/dp-cantabular-filter-flex-api/tree/develop/service) for an example of another service using this approach.
- remove the initialiser struct and check for nil pointers instead of flags, when required
- implemented post /feedback with input validation, including optional and required parametrs
- basic auth, by checking the `Authorization` header token matches the expected token in config.
- send email on valid feedback
- String sanitization:
  - HTML / XSS injection prevention. Used builtin html go library to escape string
  - SQL injection prevention. Followed [this approach](https://stackoverflow.com/questions/31647406/mysql-real-escape-string-equivalent-for-golang) to escape string
  - NoSQL injection prevention: only approached mongoDB, and escaped character '$' (we might want to discuss this approach further)

### How to review

- Make sure code changes make sense
- Make sure all tests pass
- Make sure the requirements described in the Trello card are correctly implemented

- (optional) You may run the service:
  - `make debug`
  - validate the `/health` endpoints has no checkers and 200 OK
  - validate the `POST /feedback` endpoint complies with the requirements

### Who can review

Anyone
